vowpalwabbit/cb_to_cb_adf.cc

```diff
@@ -0,0 +1,200 @@
+// Copyright (c) by respective owners including Yahoo!, Microsoft, and
+// individual contributors. All rights reserved. Released under a BSD (revised)
+// license as described in the file LICENSE.
+
+#include "cb_to_cb_adf.h"
+
+#include "vw_versions.h"
+#include "reductions.h"
+#include "learner.h"
+#include "vw.h"
+#include "cbify.h"
+#include "cb_label_parser.h"
+
+using namespace VW::LEARNER;
+using namespace VW::config;
+
+struct cb_to_cb_adf
+{
+  parameters* weights;
+  cbify_adf_data adf_data;
+  bool explore_mode;
+  multi_learner* adf_learner;
+};
+
+CB::label* get_label(cb_to_cb_adf& data, example& ec)
+{
+  if (!CB::is_test_label(ec.l.cb))
+  {
+    uint32_t chosen_action = ec.l.cb.costs[0].action - 1;
+    if (chosen_action < data.adf_data.num_actions) { return &data.adf_data.ecs[chosen_action]->l.cb; }
+  }
+
+  return nullptr;
+}
+
+template <bool is_learn>
+void predict_or_learn(cb_to_cb_adf& data, multi_learner& base, example& ec)
+{
+  data.adf_data.copy_example_to_adf(*data.weights, ec);
+
+  auto ld = get_label(data, ec);
+  CB::label saved_ld;
+
+  if (ld != nullptr)
+  {
+    saved_ld = *ld;
+    *ld = ec.l.cb;
+  }
+
+  auto restore_guard = VW::scope_exit([&ld, &saved_ld] {
+    if (ld != nullptr) *ld = saved_ld;
+  });
+
+  if (!base.learn_returns_prediction || !is_learn) { base.predict(data.adf_data.ecs); }
+
+  if (is_learn) { base.learn(data.adf_data.ecs); }
+
+  if (data.explore_mode) { ec.pred.a_s = std::move(data.adf_data.ecs[0]->pred.a_s); }
+  else
+  {
+    // cb_adf => first action is a greedy action TODO: is this a contract?
+    ec.pred.multiclass = data.adf_data.ecs[0]->pred.a_s[0].action + 1;
+  }
+}
+
+void finish_example(vw& all, cb_to_cb_adf& c, example& ec)
+{
+  if (c.explore_mode) { c.adf_data.ecs[0]->pred.a_s = std::move(ec.pred.a_s); }
+
+  auto ld = get_label(c, ec);
+  if (ld != nullptr)
+  {
+    auto saved_ld = *ld;
+    *ld = ec.l.cb;
+
+    auto restore_guard = VW::scope_exit([&ld, &saved_ld] { *ld = saved_ld; });
+
+    c.adf_learner->print_example(all, c.adf_data.ecs);
```

Comment:
is `adf_learner` used only to call `print_example`? (relevant I guess with my comment on bag print)

Reply:
yes